### PR TITLE
Explode mesh generator

### LINKED ISSUE #21060: Mesh explosion
## Reason

I want to add cohesive elements everywhere!

## Design

Add a mesh generator that breaks all element-element interfaces within given subdomains.

## Impact

This will make doing CZM/DG easier.

### PR DESCRIPTION
**Pull request by hugary1995**
Add an `ExplodeMeshGenerator` that breaks all element-element interfaces in given subdomains.

closes #21060

### CODE REVIEW

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: framework/src/meshgenerators/ExplodeMeshGenerator.C [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+  createInterface(*mesh, node_to_elem_map);
+
+  Partitioner::set_node_processor_ids(*mesh);
+
+  return dynamic_pointer_cast<MeshBase>(mesh);
+}
+
+std::unordered_map<dof_id_type, std::vector<dof_id_type>>
+ExplodeMeshGenerator::buildSubdomainRestrictedNodeToElemMap(
```
[the remark this answers is not in the document]
Comment (hugary1995) [reply framework/src/meshgenerators/ExplodeMeshGenerator.C:59]: Yeah, you are right. I can change it.

File: framework/src/meshgenerators/ExplodeMeshGenerator.C [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+void
+ExplodeMeshGenerator::duplicateNodes(
+    std::unique_ptr<MeshBase> & mesh,
+    const std::unordered_map<dof_id_type, std::vector<dof_id_type>> & node_to_elem_map) const
+{
+  for (const auto & [node_id, connected_elem_ids] : node_to_elem_map)
+    for (auto i : make_range(connected_elem_ids.size() - 1))
+      duplicateNode(mesh, mesh->elem_ptr(connected_elem_ids[i]), mesh->node_ptr(node_id));
```
[the remark this answers is not in the document]
Comment (hugary1995) [reply framework/src/meshgenerators/ExplodeMeshGenerator.C:100]: :+1:

File: framework/src/meshgenerators/ExplodeMeshGenerator.C [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  Node * new_node = nullptr;
+  new_node = Node::build(*node, mesh->n_nodes()).release();
+  new_node->processor_id() = elem->processor_id();
+  mesh->add_node(new_node);
+  for (auto j : make_range(elem->n_nodes()))
+    if (elem->node_id(j) == node->id())
+    {
+      elem->set_node(j) = new_node;
+      break;
+    }
```
[the remark this answers is not in the document]
Comment (hugary1995) [reply framework/src/meshgenerators/ExplodeMeshGenerator.C:117]: > Few things here:
> 
> let libmesh handle the ID
> don't use raw pointers for the built objects

:+1: 

> Also; is it necessary to set the processor ID?

I don't know. I was following `BreakMeshByBlockGenerator` where it assigned processor IDs. They better had a reason for that...

File: framework/src/meshgenerators/ExplodeMeshGenerator.C [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+  // Add boundary info to the new node
+  BoundaryInfo & boundary_info = mesh->get_boundary_info();
+  std::vector<boundary_id_type> node_boundary_ids;
+  boundary_info.boundary_ids(node, node_boundary_ids);
+  boundary_info.add_node(new_node, node_boundary_ids);
```
[the remark this answers is not in the document]
Comment (hugary1995) [reply framework/src/meshgenerators/ExplodeMeshGenerator.C:123]: :+1:

**Review by loganharbour — APPROVED** [review]

**Pull request merged**